### PR TITLE
[Cherry-pick] cherry pick "[MoE][PyTorch] Add mask-based MoE permutation"

### PR DESCRIPTION
**Pull request by RuibinCheung**
# Description

cherry pick "[MoE][PyTorch] Add mask-based MoE permutation" from NVIDIA upstream code.

Fixes # (issue)

## Type of change

- [x] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/triton/permutation.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    stride_probs_token,
+    stride_probs_expert,
+    # metas
+    WITH_PROBS: tl.constexpr,
+    FP8_DTYPE: tl.constexpr,
+    BLOCK_SIZE: tl.constexpr,
+):
+    if FP8_DTYPE == "e5m2":
+        compute_type = tl.float16
+        data_type = tl.float8e5
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/triton/permutation.py]: It is tl.float8e5b16 on MI300

File: transformer_engine/pytorch/triton/permutation.py [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+        pytorch_tensor_dtype = tl.uint8
+    elif FP8_DTYPE == "e4m3":
+        compute_type = tl.float16
+        data_type = tl.float8e4nv
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/triton/permutation.py]: It is tl.float8e4b8 on MI300

File: transformer_engine/pytorch/triton/permutation.py [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    inp: torch.Tensor,
+    row_id_map: torch.Tensor,
+    probs: Union[torch.Tensor, None],
+    num_tokens: int,
+    num_experts: int,
+    hidden_size: int,
+    fp8_dtype: TE_DType,
+):
+    # pylint: disable=missing-function-docstring
+    if fp8_dtype == TE_DType.kFloat8E5M2:
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/triton/permutation.py:274]: Let's customize based on the architectures here:
On MI300, use "e5m2fnuz" and "e4m3fnuz"
On NVGPU, use "e4m3fn" and "e5m2".

Mappings:
e5m2fnuz -> tl.float8e5b16  
e4m3fnuz -> tl.float8e4b8
e5m2 -> tl.float8e5
e4m3fn -> tl.float8e4nv

**Review by wenchenvincent — CHANGES_REQUESTED** [review]
Please change the fp8 data types.

File: transformer_engine/pytorch/triton/permutation.py (hunk 3 above)
Comment (RuibinCheung) [reply transformer_engine/pytorch/triton/permutation.py:274]: Acknowledge. Thanks a lot !

Comment (wenchenvincent) [reply transformer_engine/pytorch/triton/permutation.py:274]: For now, you can just assume we use "e5m2fnuz" and "e4m3fnuz" on AMDGPU (mainly MI300 family). We will refactor later when we support more ASICs which have different fp8 data formats.

**Comment by RuibinCheung** [comment]
> Please change the fp8 data types.

Done.

File: transformer_engine/pytorch/triton/permutation.py [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+        compute_type = tl.float16
+        data_type = tl.float8e5b16
+        pytorch_tensor_dtype = tl.uint8
+    elif FP8_DTYPE == "e4m3":
+        compute_type = tl.float16
+        data_type = tl.float8e4b8
+        pytorch_tensor_dtype = tl.uint8
+    else:
+        # NOTE: Using fp32 accumulate precision on ROCm.
+        compute_type = tl.float32
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/triton/permutation.py:232]: Did you see a need for this?
Usually Triton will upcast automatically if there a need.

File: transformer_engine/pytorch/triton/permutation.py (hunk 3 above)
Comment (wenchenvincent) [reply transformer_engine/pytorch/triton/permutation.py:274]: We usually want the TE code to be able to run NVGPU as well. Could you change the code here to
- use "e5m2fnuz" and "e4m3fnuz" if on AMDGPU?
- use "e4m3fn" and "e5m2" on NVGPU?

Comment (wenchenvincent) [reply transformer_engine/pytorch/triton/permutation.py:274]: from torch.utils.cpp_extension import IS_HIP_EXTENSION

IS_HIP_EXTENSION would tell you if pytorch is built for ROCm.

Comment (RuibinCheung) [reply transformer_engine/pytorch/triton/permutation.py:274]: ACK. Thx.

File: transformer_engine/pytorch/triton/permutation.py (hunk 4 above)
Comment (RuibinCheung) [reply transformer_engine/pytorch/triton/permutation.py:232]: Yes. The precision check of bwd kernel on unit test will be failed if not set compute_type to tl.float32. I suspect there are some difference on Triton backend between AMDGPU and NV GPU.

Comment (RuibinCheung) [reply transformer_engine/pytorch/triton/permutation.py:232]: I notice that NV TE upstream have changed the compute type and set it to fp32. https://github.com/NVIDIA/TransformerEngine/blob/f3a009da7b577655fe3564982d3990b06aca53b1/transformer_engine/pytorch/triton/permutation.py#L243-L252

**Review by wenchenvincent — APPROVED** [review]
LGTM.
